NuRadioReco/modules/io/RNO_G/readRNOGDataMattak.py

```diff
@@ -0,0 +1,642 @@
+                           
+         station.add_channel(channel)
+      
+      evt.set_station(station)
+      
+      return evt     
+
+
+   @register_run()
+   def run(self):
+      """ 
+      Loop over all events.
+      
+      Returns
+      -------
+      
+      evt: generator(NuRadioReco.framework.event)
+      """
+      event_idx = -1
+      for dataset in self._datasets:
+         dataset.setEntries((0, dataset.N()))
+                  
+         # read all event infos of the entier dataset (= run)
+         event_infos = dataset.eventInfo()
+         wfs = None
+
+         for idx, evtinfo in enumerate(event_infos):  # returns a list
+            event_idx += 1
+            
+            self.logger.debug(f"Processing event number {event_idx} out of total {self._n_events_total}")
+            t0 = time.time()
+                  
+            if self._filter_event(evtinfo, event_idx):
+               continue
+            
+            # Just read wfs if necessary
+            if wfs is None:
+               wfs = dataset.wfs()
+               
+            waveforms_of_event = wfs[idx]
+            
+            evt = self._get_event(evtinfo, waveforms_of_event)
+            
+            self._time_run += time.time() - t0
+            self.__counter += 1
+                           
+            yield evt
+
+
+
+   def get_event_by_index(self, event_index):
+      """ Allows to read a specific event identifed by its index 
+      
+      Parameters
+      ----------
+      
+      event_index: int
+         The index of a particluar event. The index is the chronological number from 0 to 
+         number of total events (across all datasets).
+         
+      Returns
+      -------
+      
+      evt: NuRadioReco.framework.event
+      """
+      
+      self.logger.debug(f"Processing event number {event_index} out of total {self._n_events_total}")
+      t0 = time.time()
+
+      dataset = self.__get_dataset_for_event(event_index)
+      event_info = dataset.eventInfo()  # returns a single eventInfo
+
+      if self._filter_event(event_info, event_index):
+         return None
+            
+      # access data
+      waveforms = dataset.wfs()
+      
+      evt = self._get_event(event_info, waveforms)
+      
+      self._time_run += time.time() - t0
+      self.__counter += 1
+   
+      return evt
+   
+      
+   def get_event(self, event_id):
+      """ Allows to read a specific event identifed by its id
+
+      Parameters
+      ----------
+
+      event_id: int
+         Event Id
+         
+      Returns
+      -------
+
+      evt: NuRadioReco.framework.event
+      """
+
+      self.logger.debug(f"Processing event {event_id}")
+      t0 = time.time()
+
+      event_infos = self.get_events_information(keys=["eventNumber"])
+      event_idx_ids = np.array([[index, ele["eventNumber"]] for index, ele in event_infos.items()])
+      mask = event_idx_ids[:, 1] == event_id
```

Comment:
Is it worth caching this (e.g. storing the output of get_events_information in a class variable)? Probably reading this information only is quick enough that it doesn't really matter.

Reply:
Yes it is worth it. Nope takes quite some time.